qutebrowser/browser/browsertab.py

```diff
@@ -1041,9 +1052,83 @@ def _update_load_status(self, ok: bool) -> None:
             self._set_load_status(usertypes.LoadStatus.error)
 
     @pyqtSlot()
-    def _on_history_trigger(self) -> None:
-        """Emit history_item_triggered based on backend-specific signal."""
-        raise NotImplementedError
+    def _on_history_trigger(self, force_entry: bool = True) -> None:
+        """Add or update a history entry when required."""
+        try:
+            self._widget.page()
+        except RuntimeError:
+            # Looks like this slot can be triggered on destroyed tabs:
+            # https://crashes.qutebrowser.org/view/3abffbed (Qt 5.9.1)
+            # wrapped C/C++ object of type WebEngineView has been deleted
+            log.misc.debug("Ignoring history trigger for destroyed tab")
+            return
+
+        url = self.url()
+        requested_url = self.url(requested=True)
+
+        # Don't save the title if it's generated from the URL
+        title = self.title()
+        title_url = QUrl(url)
+        title_url.setScheme('')
+        title_url_str = title_url.toDisplayString(
+            QUrl.RemoveScheme)  # type: ignore
+        if title == title_url_str.strip('/'):
+            title = ""
+
+        # Don't add history entry if the URL is invalid anyways
+        if not url.isValid():
+            log.misc.debug("Ignoring invalid URL being added to history")
+            return
+
+        if url == self._last_history_url:
+            # Same page, but title might have changed
+            if title != self._last_history_title:
+                atime = self._last_history_atime
+                # redirect=False, update=True
+                self.history_item_triggered.emit(url, title, atime,
+                                                 False, True)
+                self._last_history_title = title
+
+            # Update this in case it changed.
+            self._last_history_requested_url = requested_url
+        elif force_entry or title != self._last_history_title:
+            # Don't add an entry if only the url has changed.
```

Comment:
How about the case of navigating to two pages where the title is generated from the URL? Perhaps this should be something like `elif force_entry or not title or title != self._last_history_title:`?

Reply:
Oh, good point. I'll change that.

Reply:
On second thoughts: the title check is only for the case where the url has changed due to a javascript trigger (or something similar). Pages that are loaded 'normally' will have a new history entry generated by the trigger being called from a loadFinished signal (or the WebKit equivalent). So the only case this would affect is sites triggering Javascript/HTML5 url changes that haven't set a title, and I think it's reasonable to assume it's still the same page in this case (at least if we're doing a similar thing for an unchanged non-empty title).